libsolidity/codegen/YulUtilFunctions.cpp

```diff
@@ -1819,18 +1821,54 @@ string YulUtilFunctions::copyValueArrayStorageToStorageFunction(ArrayType const&
 				if gt(length, 0xffffffffffffffff) { <panic>() }
 				<resizeArray>(dst, length)
 
-				let srcPtr := <srcDataLocation>(src)
-
-				let dstPtr := <dstDataLocation>(dst)
+				let srcSlot := <srcDataLocation>(src)
+				let dstSlot := <dstDataLocation>(dst)
 
 				let fullSlots := div(length, <itemsPerSlot>)
-				let i := 0
-				for { } lt(i, fullSlots) { i := add(i, 1) } {
-					sstore(add(dstPtr, i), <maskFull>(sload(add(srcPtr, i))))
-				}
-				let spill := sub(length, mul(i, <itemsPerSlot>))
+				let spill := sub(length, mul(fullSlots, <itemsPerSlot>))
+				let totalSlots := fullSlots
 				if gt(spill, 0) {
-					sstore(add(dstPtr, i), <maskBytes>(sload(add(srcPtr, i)), mul(spill, <bytesPerItem>)))
+					totalSlots := add(totalSlots, 1)
+				}
+
+				let srcSlotValue := sload(srcSlot)
+				<?srcReadMultiPerSlot>let srcItemIndexInSlot := 0</srcReadMultiPerSlot>
+
+				for { let i := 0 } lt(i, totalSlots) { i := add(i, 1) } {
+					let dstSlotValue := 0
+					<?sameType>
```

Comment:
Maybe instead of distinguishing the "sameType" case, we should rather distinguish the case "srcStride" and "dstStride" are both at least 17 - i.e. we have one item per slot.

Reply:
I would stay with same type, as if types is not the same we need a call to conversion function

Reply:
Ah wait a second - the "sameType" case also works with multiple items per slot? Then it should certainly have its own case.

Reply:
Still, the "only one item per slot" could also be its own case - I don't think loop unrolling is coming any time soon...

Reply:
Although maybe it should...

Reply:
Updated. Also there are cases when types are not the same, src base type is less then 17 bytes and dst base type is more then 16 bytes and all the cases are optimized as much as possible now (and covered with tests).